src/evaluate.cpp

```diff
@@ -23,42 +23,93 @@
+#include "misc.h"
 #include "pawns.h"
 #include "thread.h"
 #include "uci.h"
+#include "incbin/incbin.h"
+
+
+
+// Macro to embed the default NNUE file data in the engine binary (using incbin.h, by Dale Weiler)
+// This macro invocation will declare the following three variables
+//     const unsigned char        gEmbededNNUEData[];  // a pointer to the embeded data
+//     const unsigned char *const gEmbededNNUEEnd;     // a marker to the end
+//     const unsigned int         gEmbededNNUESize;    // the size of the embeded file
+INCBIN(EmbededNNUE, EvalFileDefaultName);
+// [TODO] Build will fail if default net is not available in the sources.
+
+
+
+using namespace std;
+using namespace Eval::NNUE;
 
 namespace Eval {
 
   bool useNNUE;
-  std::string eval_file_loaded="None";
+  string eval_file_loaded = "None";
 
-  void init_NNUE() {
+  /// init_NNUE() tries to load a nnue network at startup time, or when the engine
+  /// receives a UCI command "setoption name EvalFile value blahblah.nnue"
+  /// The name of the nnue network is always retrieved from the EvalFile option.
+  /// We search the given network in three locations: internally (the default
+  /// network may be embeded in the binary), in the active working directory and
+  /// in the engine directory.
 
+  void init_NNUE() {
     useNNUE = Options["Use NNUE"];
-    std::string eval_file = std::string(Options["EvalFile"]);
-    if (useNNUE && eval_file_loaded != eval_file)
-        if (Eval::NNUE::load_eval_file(eval_file))
-            eval_file_loaded = eval_file;
+    string eval_file = string(Options["EvalFile"]);
+
+    vector<string> dirs = { "<internal>", "" , CommandLine::binaryDirectory };
+
+    if (useNNUE)
+        for (string directory : dirs)
+            if (eval_file_loaded != eval_file)
+            {
+                if (directory == "<internal>" && eval_file == EvalFileDefaultName)
+                {
+                    cerr << "Trying to load eval from memory... " << eval_file << endl;
+                    if (load_eval_from_memory(eval_file, (char*)gEmbededNNUEData, size_t(gEmbededNNUESize )))
+                    {
+                        eval_file_loaded = eval_file;
+                        cerr << "Eval loaded from memory: OK" << endl;
+                    }
+                }
+                else
+                {
+                    cerr << "Trying to find eval file on disc... " << directory + eval_file << endl;
+                    if (load_eval_file(directory + eval_file))
+                    {
+                        eval_file_loaded = eval_file;
+                        cerr << "Eval loaded from disc: OK" << endl;
+                    }
+                }
+            }
   }
 
+  /// verify_NNUE() verifies that the last net used was loaded successfully
   void verify_NNUE() {
 
-    std::string eval_file = std::string(Options["EvalFile"]);
+    assert( ("Verifying size of embeded NNUE file" , gEmbededNNUESize == 21022697) );
```

Comment:
I would not verify the size, it will change with different version of the net. Just let the reading / parsing of the net fail.